### ISSUE TITLE
Proposal: introduce repeatWhileActive top level suspend function

### ISSUE THREAD

**Issue by LouisCAD** [Kotlin/kotlinx.coroutines#1063]
Here's how it would be implemented:
```kotlin
import kotlin.coroutines.coroutineContext
import kotlinx.coroutines.ensureActive

suspend inline fun repeatWhileActive(block: () -> Unit): Nothing {
    while (true) {
        coroutineContext.ensureActive()
        block()
    }
}
```
This has the benefit of being an expression that can be used in lambdas that return a value (which is not the case for raw `while(true)`, because it is not considered as an expression.

I can make a PR for this.

**Comment by LouisCAD** [comment]
This would allow such things:
```kotlin
suspend fun runMyStuff(someParam: SomeType): Nothing = coroutineScope {
    repeatWhileActive {
        doStuff()
    }
}
```
while the following snippet doesn't compile:
```kotlin
suspend fun runMyStuff(someParam: SomeType): Nothing = coroutineScope {
    while (true) {
        coroutineContext.ensureActive()
        doStuff()
    } // Compilation error: while is not an expression
}
```

**Comment by elizarov** [comment]
What are use-cases?

**Comment by LouisCAD** [comment]
I mainly use this in user interfaces where I loop on some UI process that can be completed and done again (example: taking a video clip, reviewing it, make a sending request to a contact, and doing it again until the user presses the back button).

I found using coroutines in UI to be a perfect fit because thanks to structured concurrency (and `try`/`finally`), I can ensure things are actionable and visible only in a certain scope. For example, showing the hang up button and waiting for the user to click on it only while there's a call ongoing.

Since I'm currently primarily doing Android development, I made a few extensions for Android's `View` like `awaitOneClick()` that can also hide the view after it's clicked, removing any need for debounce (that can still be applied if looping by simply using `delay` at the end of the loop before the next run.
You can see the extensions here: https://github.com/LouisCAD/Splitties/blob/master/modules/views-coroutines/README.md

I unfortunately don't have an open source example of these at play at the moment, but I'm using these with `repeatWhileActive` a lot in my company's projects now.

**Comment by LouisCAD** [comment]
Here's an example of something I'm working on to show a use case:
```kotlin
override suspend fun promptItemsAvailabilityCheckAndAwaitMissingItemsIds(
    items: List<SomeItem>
): List<Int> {
    itemsAdapter.submitList(mode, items)
    repeatWhileActive {
        confirmBtn.awaitOneClick()
        TODO("Get missing items")
        TODO("If ok, return emptyList") // Possible because repeatWhileActive is inline
        TODO("else, alert and ask for confirmation")
        TODO("If confirmed, return the ids of missing items.")
        TODO("else, let the loop iterate again")
    }
}
```

**Comment by streetsofboston** [comment]
Wouldn't this be a general lambda issue in Kotlin, not just for `suspend` lambdas?

E.g. this doesn't compile either:

```
// Does not compile
val doStuffForever : () -> Nothing = {
    while (true) {
        // do stuff
    }
}

// But this does compile
fun doStuffForever() : Nothing {
    while (true) {
        // do stuff
    }
}
```

**Comment by elizarov** [comment]
@streetsofboston Yes, indeed. The problem will be some alleviated if infinite loop is introduced to the language, see https://youtrack.jetbrains.com/issue/KT-27970

**Comment by LouisCAD** [comment]
In the meantime, should I make a PR for that proposal, possibly marking it as experimental, or should I put it in a library of mine?

**Comment by elizarov** [comment]
Please, don't do PR yet, but you are definitely welcome to experiment in your own code.

**Comment by dkhalanskyjb** [comment]
See also: #1632, which eventually converged to almost the same thing.

**Comment by dkhalanskyjb** [comment]
Current state of things: this still doesn't compile
```kotlin
suspend fun runMyStuff(): Nothing = coroutineScope {
    while (true) {
        // blank
    }
}
```
However, this compiles...
```kotlin
suspend fun runMyStuff(): Nothing = coroutineScope {
    while (true) {
        // blank
    }
    throw IllegalArgumentException()
}
```
... but says that `throw IllegalArgumentException()` is unreachable! So, the compiler knows that `while (true)` never exits, but does not allow typing `while (true)` as `Nothing`.
